File: python/test/runtimecontrol_test.py
import sys, os
sys.path.insert(0, os.path.dirname(__file__) + "/..")

import unittest
import inhumate_rti as RTI
import time


def wait_for(condition, timeout=2.0):
    count = 0
    max_count = int(timeout * 100)
    while count < max_count and not condition():
        count += 1
        time.sleep(0.01)


class RuntimeControlTest(unittest.TestCase):

    @classmethod
    def setUpClass(cls):
        def on_error(channel, message, exception):
            print(f"Error: {channel}: {message}", file=sys.stderr)

        cls.rti = RTI.Client(application="python_rc_test")
        cls.rti.on("error", on_error)
        cls.rti.wait_until_connected()

        # Separate client that acts as the controller publishing control messages
        cls.controller = RTI.Client(application="python_rc_test_controller")
        cls.controller.on("error", on_error)
        cls.controller.wait_until_connected()

        cls.runtime = RTI.RuntimeControl(cls.rti)
        time.sleep(0.1)

    @classmethod
    def tearDownClass(cls):
        cls.rti.disconnect()
        cls.controller.disconnect()

    def setUp(self):
        time.sleep(0.05)  # let any previous test's messages drain
        self.rti.state = RTI.proto.INITIAL

    def _publish(self, msg):
        self.controller.publish(RTI.channel.control, msg)

    def _fresh_client(self, name="python_rc_hook_test"):
        rti = RTI.Client(application=name)
        rti.on("error", lambda c, m, e: print(f"Error: {c}: {m}", file=sys.stderr))
        rti.wait_until_connected()
        return rti

    def _configure_run(self, run_id, time_step=1.0):
        msg = RTI.proto.FastTimeControl()
        msg.configure.controller_client_id = self.controller.client_id
        msg.configure.run_id = run_id
        msg.configure.time_step = time_step
        self.controller.publish(RTI.channel.fast_time_control, msg)

    def _send_grant(self, run_id, step_number=0, start_time=0.0, time_step=1.0):
        msg = RTI.proto.FastTimeControl()
        msg.step_grant.run_id = run_id
        msg.step_grant.step_number = step_number
        msg.step_grant.start_time = start_time
        msg.step_grant.end_time = start_time + time_step
        self.controller.publish(RTI.channel.fast_time_control, msg)

    # --- Capabilities ---

    def test_capabilities_added(self):
        self.assertIn(RTI.capability.runtime_control, self.rti.capabilities)
        self.assertIn(RTI.capability.scenario, self.rti.capabilities)
        self.assertIn(RTI.capability.time_scale, self.rti.capabilities)

    def test_fast_time_capability_added(self):
        rti = self._fresh_client("python_rc_ft_cap_test")
        try:
            runtime = RTI.RuntimeControl(rti, fast_time=True)
            self.assertIn(RTI.capability.fast_time_worker, rti.capabilities)
        finally:
            rti.disconnect()

    # --- State transitions ---

    def test_initial_state_is_initial(self):
        self.assertEqual(RTI.proto.INITIAL, self.rti.state)

    def test_start_sets_state_running(self):
        msg = RTI.proto.RuntimeControl()
        msg.start.SetInParent()
        self._publish(msg)
        wait_for(lambda: self.rti.state == RTI.proto.RUNNING)
        self.assertEqual(RTI.proto.RUNNING, self.rti.state)

    def test_pause_sets_state_paused(self):
        self.rti.state = RTI.proto.RUNNING
        msg = RTI.proto.RuntimeControl()
        msg.pause.SetInParent()
        self._publish(msg)
        wait_for(lambda: self.rti.state == RTI.proto.PAUSED)
        self.assertEqual(RTI.proto.PAUSED, self.rti.state)

    def test_stop_sets_state_stopped(self):
        self.rti.state = RTI.proto.RUNNING
        msg = RTI.proto.RuntimeControl()
        msg.stop.SetInParent()
        self._publish(msg)
        wait_for(lambda: self.rti.state == RTI.proto.STOPPED)
        self.assertEqual(RTI.proto.STOPPED, self.rti.state)

    def test_end_sets_state_end(self):
        self.rti.state = RTI.proto.RUNNING
        msg = RTI.proto.RuntimeControl()
        msg.end.SetInParent()
        self._publish(msg)
        wait_for(lambda: self.rti.state == RTI.proto.END)
        self.assertEqual(RTI.proto.END, self.rti.state)

    def test_reset_sets_state_initial(self):
        self.rti.state = RTI.proto.RUNNING
        msg = RTI.proto.RuntimeControl()
        msg.reset.SetInParent()
        self._publish(msg)
        wait_for(lambda: self.rti.state == RTI.proto.INITIAL)
        self.assertEqual(RTI.proto.INITIAL, self.rti.state)

    def test_load_scenario_sets_state_ready(self):
        msg = RTI.proto.RuntimeControl()
        msg.load_scenario.name = "TestScenario"
        self._publish(msg)
        wait_for(lambda: self.rti.state == RTI.proto.READY)
        self.assertEqual(RTI.proto.READY, self.rti.state)

    def test_load_scenario_stores_scenario(self):
        msg = RTI.proto.RuntimeControl()
        msg.load_scenario.name = "UniqueScenario42"
        self._publish(msg)
        wait_for(lambda: self.runtime.scenario is not None and self.runtime.scenario.name == "UniqueScenario42")
        self.assertEqual("UniqueScenario42", self.runtime.scenario.name)

    def test_set_time_scale_updates_property(self):
        msg = RTI.proto.RuntimeControl()
        msg.set_time_scale.time_scale = 2.5
        self._publish(msg)
        wait_for(lambda: self.runtime.time_scale == 2.5)
        self.assertAlmostEqual(2.5, self.runtime.time_scale)

    def test_time_sync_updates_time_scale(self):
        msg = RTI.proto.RuntimeControl()
        msg.time_sync.time_scale = 4.0
        self._publish(msg)
        wait_for(lambda: self.runtime.time_scale == 4.0)
        self.assertAlmostEqual(4.0, self.runtime.time_scale)

    # --- Override hooks (fresh client per test for isolation) ---

    def test_on_reset_called(self):
        rti = self._fresh_client()
        try:
            called = []
            class TestRuntime(RTI.RuntimeControl):
                def on_reset(self): called.append(True)
            TestRuntime(rti)
            msg = RTI.proto.RuntimeControl()
            msg.reset.SetInParent()
            self.controller.publish(RTI.channel.control, msg)
            wait_for(lambda: len(called) > 0)
            self.assertTrue(called)
        finally:
            rti.disconnect()

    def test_on_start_called(self):
        rti = self._fresh_client()
        try:
            called = []
            class TestRuntime(RTI.RuntimeControl):
                def on_start(self): called.append(True)
            TestRuntime(rti)
            msg = RTI.proto.RuntimeControl()
            msg.start.SetInParent()
            self.controller.publish(RTI.channel.control, msg)
            wait_for(lambda: len(called) > 0)
            self.assertTrue(called)
        finally:
            rti.disconnect()

    def test_on_stop_called(self):
        rti = self._fresh_client()
        try:
            called = []
            class TestRuntime(RTI.RuntimeControl):
                def on_stop(self): called.append(True)
            TestRuntime(rti)
            msg = RTI.proto.RuntimeControl()
            msg.stop.SetInParent()
            self.controller.publish(RTI.channel.control, msg)
            wait_for(lambda: len(called) > 0)
            self.assertTrue(called)
        finally:
            rti.disconnect()

    def test_on_load_scenario_false_sets_unknown(self):
        rti = self._fresh_client()
        try:
            class TestRuntime(RTI.RuntimeControl):
                def on_load_scenario(self, load_scenario, playback): return False
            TestRuntime(rti)
            msg = RTI.proto.RuntimeControl()
            msg.load_scenario.name = "Nope"
            self.controller.publish(RTI.channel.control, msg)
            wait_for(lambda: rti.state == RTI.proto.UNKNOWN)
            self.assertEqual(RTI.proto.UNKNOWN, rti.state)
        finally:
            rti.disconnect()

    # --- Fast-time (fresh client per test) ---

    def test_fast_time_configure_sends_acknowledge(self):
        rti = self._fresh_client("python_rc_ft_ack_test")
        acked = []
        sub = self.controller.subscribe(RTI.channel.fast_time_control, RTI.proto.FastTimeControl,
            lambda c, m: acked.append(m.acknowledge) if m.HasField("acknowledge") else None)
        try:
            RTI.RuntimeControl(rti, fast_time=True)
            time.sleep(0.5)
            self._configure_run("run-ack")
            wait_for(lambda: any(a.client_id == rti.client_id for a in acked))
            self.assertTrue(any(a.client_id == rti.client_id for a in acked))
            mine = next(a for a in acked if a.client_id == rti.client_id)
            self.assertEqual("run-ack", mine.run_id)
        finally:
            self.controller.unsubscribe(sub)
            rti.disconnect()

    def test_fast_time_is_fast_time_after_configure(self):
        rti = self._fresh_client("python_rc_ft_isft_test")
        try:
            runtime = RTI.RuntimeControl(rti, fast_time=True)
            self.assertFalse(runtime.is_fast_time)
            time.sleep(0.5)
            self._configure_run("run-isft")
            wait_for(lambda: runtime.is_fast_time)
            self.assertTrue(runtime.is_fast_time)
        finally:
            rti.disconnect()

    def test_fast_time_dispatch_mode_buffered_after_configure(self):
        rti = self._fresh_client("python_rc_ft_disp_test")
        try:
            RTI.RuntimeControl(rti, fast_time=True)
            self.assertEqual(RTI.DispatchMode.IMMEDIATE, rti.default_dispatch_mode)
            time.sleep(0.5)
            self._configure_run("run-dispatch")
            wait_for(lambda: rti.default_dispatch_mode == RTI.DispatchMode.BUFFERED)
            self.assertEqual(RTI.DispatchMode.BUFFERED, rti.default_dispatch_mode)
        finally:
            rti.disconnect()

    def test_fast_time_step_fn_called_and_completes(self):
        rti = self._fresh_client("python_rc_ft_step_test")
        steps = []
        completions = []
        sub = self.controller.subscribe(RTI.channel.fast_time_control, RTI.proto.FastTimeControl,
            lambda c, m: completions.append(m.step_complete) if m.HasField("step_complete") else None)
        try:
            RTI.RuntimeControl(rti, step_fn=lambda g: steps.append(g))
            time.sleep(0.5)
            self._configure_run("run-step-fn", time_step=0.5)
            time.sleep(0.1)
            self._send_grant("run-step-fn", time_step=0.5)
            wait_for(lambda: any(c.client_id == rti.client_id for c in completions) and len(steps) > 0)
            self.assertTrue(len(steps) > 0)
            self.assertAlmostEqual(0.5, steps[0].time_step)
            mine = next(c for c in completions if c.client_id == rti.client_id)
            self.assertFalse(mine.failed)
            self.assertEqual(0, mine.step_number)
        finally:
            self.controller.unsubscribe(sub)
            rti.disconnect()

    def test_fast_time_wait_for_step_grant(self):
        rti = self._fresh_client("python_rc_ft_wait_test")
        try:
            runtime = RTI.RuntimeControl(rti, fast_time=True)
            time.sleep(0.5)
            self._configure_run("run-wait", time_step=1.0)
            time.sleep(0.1)
            self._send_grant("run-wait", time_step=1.0)
            grant = runtime.wait_for_step_grant(timeout=2.0)
            self.assertIsNotNone(grant)
            self.assertAlmostEqual(1.0, grant.time_step)
            self.assertEqual(0, grant.step_number)
            runtime.complete_step(grant)
        finally:
            rti.disconnect()

    def test_fast_time_stop_clears_is_fast_time(self):
        rti = self._fresh_client("python_rc_ft_stop_test")
        try:
            runtime = RTI.RuntimeControl(rti, fast_time=True)
            time.sleep(0.5)
            self._configure_run("run-stop")
            wait_for(lambda: runtime.is_fast_time)
            self.assertTrue(runtime.is_fast_time)
            msg = RTI.proto.RuntimeControl()
            msg.stop.SetInParent()
            self.controller.publish(RTI.channel.control, msg)
            wait_for(lambda: not runtime.is_fast_time)
            self.assertFalse(runtime.is_fast_time)
        finally:
            rti.disconnect()

    def test_fast_time_wait_for_step_grant_returns_none_on_stop(self):
        rti = self._fresh_client("python_rc_ft_none_test")
        try:
            runtime = RTI.RuntimeControl(rti, fast_time=True)
            time.sleep(0.5)
            self._configure_run("run-none")
            wait_for(lambda: runtime.is_fast_time)
            # Stop while waiting — WaitForStepGrant must unblock and return None
            def do_stop():
                time.sleep(0.1)
                msg = RTI.proto.RuntimeControl()
                msg.stop.SetInParent()
                self.controller.publish(RTI.channel.control, msg)
            import threading
            threading.Thread(target=do_stop, daemon=True).start()
            grant = runtime.wait_for_step_grant(timeout=2.0)
            self.assertIsNone(grant)
        finally:
            rti.disconnect()


if __name__ == "__main__":
    unittest.main()

File: dotnet/test/RuntimeControlTest.cs
using NUnit.Framework;
using System;
using System.Threading;
using System.Diagnostics;
using Inhumate.RTI.Proto;

namespace Inhumate.RTI {
    [NonParallelizable]
    public class RuntimeControlTest {

        protected static RTIClient rti;
        protected static RTIClient controller;
        protected static RTIRuntimeControl runtime;

        private static int TestTimeoutMs =>
            int.TryParse(Environment.GetEnvironmentVariable("RTI_TEST_TIMEOUT_MS"), out var timeoutMs)
                ? timeoutMs
                : Environment.GetEnvironmentVariable("GITLAB_CI") == "true" ? 60000 : 30000;

        [OneTimeSetUp]
        public static void Setup() {
            rti = FreshClient("C# RuntimeControlTest");

            // Separate client that acts as the controller publishing control messages
            controller = FreshClient("C# RuntimeControlTest Controller");

            runtime = new RTIRuntimeControl(rti);
            Thread.Sleep(100);
        }

        [OneTimeTearDown]
        public static void Teardown() {
            rti?.Disconnect();
            controller?.Disconnect();
            Thread.Sleep(100);
        }

        [SetUp]
        public void ResetState() {
            Thread.Sleep(50); // let any previous test's messages drain
            rti.State = RuntimeState.Initial;
        }

        private static void Publish(RuntimeControl message) =>
            controller.Publish(RTIChannel.RuntimeControl, message);

        private static RTIClient FreshClient(string name = "C# RuntimeControlTest Fresh") {
            Exception lastException = null;
            var sw = Stopwatch.StartNew();
            while (sw.ElapsedMilliseconds < TestTimeoutMs) {
                var c = new RTIClient(connect: false) { Application = name };
                c.OnError += (ch, ex) => Console.Error.WriteLine($"Error: {ch}: {ex}");
                try {
                    c.Connect();
                    c.WaitUntilConnected(Math.Max(1000, (int)Math.Min(5000, TestTimeoutMs - sw.ElapsedMilliseconds)));
                    return c;
                } catch (Exception ex) {
                    lastException = ex;
                    try { c.Disconnect(); } catch (Exception) { }
                    Thread.Sleep(250);
                }
            }
            throw new RTIConnectionFailure($"Could not connect {name} within {TestTimeoutMs} ms: {lastException?.Message}");
        }

        private void ConfigureRun(string runId, double timeStep = 1.0) =>
            controller.Publish(RTIChannel.FastTimeControl, new FastTimeControl {
                Configure = new FastTimeControl.Types.Configure {
                    ControllerClientId = controller.ClientId,
                    RunId = runId,
                    TimeStep = timeStep
                }
            });

        private void SendGrant(string runId, long stepNumber = 0, double startTime = 0.0, double timeStep = 1.0) =>
            controller.Publish(RTIChannel.FastTimeControl, new FastTimeControl {
                StepGrant = new FastTimeControl.Types.StepGrant {
                    RunId = runId,
                    StepNumber = stepNumber,
                    StartTime = startTime,
                    EndTime = startTime + timeStep
                }
            });

        private static bool WaitFor(Func<bool> condition, int? timeoutMs = null) {
            var timeout = timeoutMs ?? TestTimeoutMs;
            var sw = Stopwatch.StartNew();
            while (!condition() && sw.ElapsedMilliseconds < timeout) Thread.Sleep(10);
            return condition();
        }

        // --- Capabilities ---

        [Test]
        public void Capabilities_Added() {
            Assert.IsTrue(rti.Capabilities.Contains(RTICapability.RuntimeControl));
            Assert.IsTrue(rti.Capabilities.Contains(RTICapability.Scenario));
            Assert.IsTrue(rti.Capabilities.Contains(RTICapability.TimeScale));
        }

        [Test]
        public void FastTime_Capability_Added() {
            var c = FreshClient("C# RC FastTime Cap Test");
            try {
                new RTIRuntimeControl(c, fastTime: true);
                Assert.IsTrue(c.Capabilities.Contains(RTICapability.FastTimeWorker));
            } finally {
                c.Disconnect();
            }
        }

        // --- State transitions ---

        [Test]
        public void Initial_State_Is_Initial() {
            Assert.AreEqual(RuntimeState.Initial, rti.State);
        }

        [Test]
        public void Start_SetsState_Running() {
            Publish(new RuntimeControl { Start = new Google.Protobuf.WellKnownTypes.Empty() });
            Assert.IsTrue(WaitFor(() => rti.State == RuntimeState.Running));
            Assert.AreEqual(RuntimeState.Running, rti.State);
        }

        [Test]
        public void Pause_SetsState_Paused() {
            rti.State = RuntimeState.Running;
            Publish(new RuntimeControl { Pause = new Google.Protobuf.WellKnownTypes.Empty() });
            Assert.IsTrue(WaitFor(() => rti.State == RuntimeState.Paused));
            Assert.AreEqual(RuntimeState.Paused, rti.State);
        }

        [Test]
        public void Stop_SetsState_Stopped() {
            rti.State = RuntimeState.Running;
            Publish(new RuntimeControl { Stop = new Google.Protobuf.WellKnownTypes.Empty() });
            Assert.IsTrue(WaitFor(() => rti.State == RuntimeState.Stopped));
            Assert.AreEqual(RuntimeState.Stopped, rti.State);
        }

        [Test]
        public void End_SetsState_End() {
            rti.State = RuntimeState.Running;
            Publish(new RuntimeControl { End = new Google.Protobuf.WellKnownTypes.Empty() });
            Assert.IsTrue(WaitFor(() => rti.State == RuntimeState.End));
            Assert.AreEqual(RuntimeState.End, rti.State);
        }

        [Test]
        public void Reset_SetsState_Initial() {
            rti.State = RuntimeState.Running;
            Publish(new RuntimeControl { Reset = new Google.Protobuf.WellKnownTypes.Empty() });
            Assert.IsTrue(WaitFor(() => rti.State == RuntimeState.Initial));
            Assert.AreEqual(RuntimeState.Initial, rti.State);
        }

        [Test]
        public void LoadScenario_SetsState_Ready() {
            Publish(new RuntimeControl { LoadScenario = new RuntimeControl.Types.ScenarioSpecification { Name = "TestScene" } });
            Assert.IsTrue(WaitFor(() => rti.State == RuntimeState.Ready));
            Assert.AreEqual(RuntimeState.Ready, rti.State);
        }

        [Test]
        public void LoadScenario_StoresScenario() {
            Publish(new RuntimeControl { LoadScenario = new RuntimeControl.Types.ScenarioSpecification { Name = "UniqueScene99" } });
            // Wait for State==Ready (set after Scenario, and involves PublishClient memory barrier)
            // before reading Scenario to avoid ARM memory-ordering issues.
            Assert.IsTrue(WaitFor(() => rti.State == RuntimeState.Ready));
            Assert.AreEqual("UniqueScene99", runtime.Scenario?.Name);
        }

        [Test]
        public void SetTimeScale_UpdatesProperty() {
            Publish(new RuntimeControl { SetTimeScale = new RuntimeControl.Types.SetTimeScale { TimeScale = 2.5 } });
            Assert.IsTrue(WaitFor(() => runtime.TimeScale == 2.5));
            Assert.AreEqual(2.5, runtime.TimeScale.Value, 0.001);
        }

        [Test]
        public void TimeSync_UpdatesTimeScale() {
            Publish(new RuntimeControl { TimeSync = new RuntimeControl.Types.TimeSync { TimeScale = 4.0 } });
            Assert.IsTrue(WaitFor(() => runtime.TimeScale == 4.0));
            Assert.AreEqual(4.0, runtime.TimeScale.Value, 0.001);
        }

        [Test]
        public void Seek_FromInitial_SetsPlaybackPaused() {
            rti.State = RuntimeState.Initial;
            Publish(new RuntimeControl { Seek = new RuntimeControl.Types.Seek { Time = 12.5 } });
            Assert.IsTrue(WaitFor(() => rti.State == RuntimeState.PlaybackPaused));
            Assert.AreEqual(RuntimeState.PlaybackPaused, rti.State);
        }

        [Test]
        public void Seek_DuringPlayback_LeavesStateUntouched() {
            rti.State = RuntimeState.Playback;
            Publish(new RuntimeControl { Seek = new RuntimeControl.Types.Seek { Time = 1.0 } });
            Thread.Sleep(100);
            Assert.AreEqual(RuntimeState.Playback, rti.State);
        }

        [Test]
        public void Seek_DuringRunning_LeavesStateUntouched() {
            rti.State = RuntimeState.Running;
            Publish(new RuntimeControl { Seek = new RuntimeControl.Types.Seek { Time = 1.0 } });
            Thread.Sleep(100);
            Assert.AreEqual(RuntimeState.Running, rti.State);
        }

        [Test]
        public void Seek_DuringPaused_LeavesStateUntouched() {
            rti.State = RuntimeState.Paused;
            Publish(new RuntimeControl { Seek = new RuntimeControl.Types.Seek { Time = 1.0 } });
            Thread.Sleep(100);
            Assert.AreEqual(RuntimeState.Paused, rti.State);
        }

        [Test]
        public void Seek_DuringPlaybackStopped_SetsPlaybackPaused() {
            rti.State = RuntimeState.PlaybackStopped;
            Publish(new RuntimeControl { Seek = new RuntimeControl.Types.Seek { Time = 1.0 } });
            Assert.IsTrue(WaitFor(() => rti.State == RuntimeState.PlaybackPaused));
            Assert.AreEqual(RuntimeState.PlaybackPaused, rti.State);
        }

        [Test]
        public void OnSeek_Called() {
            double receivedTime = -1;
            var c = FreshClient("C# RC OnSeek Test");
            try {
                new TestRuntime(c, onSeek: s => { receivedTime = s.Time; });
                Thread.Sleep(250);
                controller.Publish(RTIChannel.RuntimeControl, new RuntimeControl { Seek = new RuntimeControl.Types.Seek { Time = 7.25 } });
                Assert.IsTrue(WaitFor(() => receivedTime > 0));
                Assert.AreEqual(7.25, receivedTime, 0.001);
            } finally {
                c.Disconnect();
            }
        }

        // --- Override hooks (fresh client per test for isolation) ---

        [Test]
        public void OnReset_Called() {
            bool called = false;
            var c = FreshClient("C# RC OnReset Test");
            try {
                new TestRuntime(c, onReset: () => { called = true; });
                Thread.Sleep(250);
                controller.Publish(RTIChannel.RuntimeControl, new RuntimeControl { Reset = new Google.Protobuf.WellKnownTypes.Empty() });
                Assert.IsTrue(WaitFor(() => called));
            } finally {
                c.Disconnect();
            }
        }

        [Test]
        public void OnStart_Called() {
            bool called = false;
            var c = FreshClient("C# RC OnStart Test");
            try {
                new TestRuntime(c, onStart: () => { called = true; });
                Thread.Sleep(250);
                controller.Publish(RTIChannel.RuntimeControl, new RuntimeControl { Start = new Google.Protobuf.WellKnownTypes.Empty() });
                Assert.IsTrue(WaitFor(() => called));
            } finally {
                c.Disconnect();
            }
        }

        [Test]
        public void OnStop_Called() {
            bool called = false;
            var c = FreshClient("C# RC OnStop Test");
            try {
                new TestRuntime(c, onStop: () => { called = true; });
                Thread.Sleep(250);
                controller.Publish(RTIChannel.RuntimeControl, new RuntimeControl { Stop = new Google.Protobuf.WellKnownTypes.Empty() });
                Assert.IsTrue(WaitFor(() => called));
            } finally {
                c.Disconnect();
            }
        }

        [Test]
        public void OnLoadScenario_ReturnsFalse_SetsUnknown() {
            var c = FreshClient("C# RC LoadFalse Test");
            try {
                new TestRuntime(c, loadScenarioResult: false);
                Thread.Sleep(250);
                controller.Publish(RTIChannel.RuntimeControl, new RuntimeControl { LoadScenario = new RuntimeControl.Types.ScenarioSpecification { Name = "Nope" } });
                Assert.IsTrue(WaitFor(() => c.State == RuntimeState.Unknown));
                Assert.AreEqual(RuntimeState.Unknown, c.State);
            } finally {
                c.Disconnect();
            }
        }

        // --- Fast-time (fresh client per test) ---

        [Test]
        public void FastTime_Configure_SendsAcknowledge() {
            var c = FreshClient("C# RC FT Ack Test");
            FastTimeControl.Types.Acknowledge acked = null;
            var sub = controller.Subscribe<FastTimeControl>(RTIChannel.FastTimeControl, (ch, msg) => {
                if (msg.ControlCase == FastTimeControl.ControlOneofCase.Acknowledge && msg.Acknowledge.RunId == "run-ack" && msg.Acknowledge.ClientId == c.ClientId)
                    acked = msg.Acknowledge;
            });
            try {
                new RTIRuntimeControl(c, fastTime: true);
                Thread.Sleep(250);
                ConfigureRun("run-ack");
                Assert.IsTrue(WaitFor(() => acked != null));
                Assert.AreEqual("run-ack", acked.RunId);
            } finally {
                controller.Unsubscribe(sub);
                c.Disconnect();
            }
        }

        [Test]
        public void FastTime_IsFastTime_AfterConfigure() {
            var c = FreshClient("C# RC FT Isft Test");
            try {
                var rt = new RTIRuntimeControl(c, fastTime: true);
                Assert.IsFalse(rt.IsFastTime);
                Thread.Sleep(100);
                ConfigureRun("run-isft");
                Assert.IsTrue(WaitFor(() => rt.IsFastTime));
                Assert.IsTrue(rt.IsFastTime);
            } finally {
                c.Disconnect();
            }
        }

        [Test]
        public void FastTime_DispatchMode_Immediate_AfterConfigure_Buffered_AfterStep() {
            var c = FreshClient("C# RC FT Dispatch Test");
            try {
                var rt = new RTIRuntimeControl(c, fastTime: true);
                Assert.AreEqual(DispatchMode.Immediate, c.DefaultDispatchMode);
                Thread.Sleep(100);
                ConfigureRun("run-dispatch");
                Assert.IsTrue(WaitFor(() => rt.IsFastTime));
                // Still Immediate after configure — only switches on first step grant
                Assert.AreEqual(DispatchMode.Immediate, c.DefaultDispatchMode);
                SendGrant("run-dispatch", timeStep: 1.0);
                Assert.IsTrue(WaitFor(() => c.DefaultDispatchMode == DispatchMode.Buffered));
                Assert.AreEqual(DispatchMode.Buffered, c.DefaultDispatchMode);
            } finally {
                c.Disconnect();
            }
        }

        [Test]
        public void FastTime_Play_ResetsFastTime() {
            var c = FreshClient("C# RC FT Play Test");
            try {
                var rt = new RTIRuntimeControl(c, fastTime: true);
                Thread.Sleep(100);
                ConfigureRun("run-play");
                Assert.IsTrue(WaitFor(() => rt.IsFastTime));
                controller.Publish(RTIChannel.RuntimeControl, new RuntimeControl { Play = new Google.Protobuf.WellKnownTypes.Empty() });
                Assert.IsTrue(WaitFor(() => !rt.IsFastTime));
                Assert.IsFalse(rt.IsFastTime);
                Assert.AreEqual(DispatchMode.Immediate, c.DefaultDispatchMode);
            } finally {
                c.Disconnect();
            }
        }

        [Test]
        public void FastTime_StepFn_CalledAndCompletes() {
            var c = FreshClient("C# RC FT StepFn Test");
            StepGrant receivedGrant = null;
            FastTimeControl.Types.StepComplete completion = null;
            var sub = controller.Subscribe<FastTimeControl>(RTIChannel.FastTimeControl, (ch, msg) => {
                if (msg.ControlCase == FastTimeControl.ControlOneofCase.StepComplete && msg.StepComplete.RunId == "run-step-fn")
                    completion = msg.StepComplete;
            });
            try {
                var rt = new RTIRuntimeControl(c, stepFn: g => { receivedGrant = g; });
                Thread.Sleep(250);
                ConfigureRun("run-step-fn", timeStep: 0.5);
                Assert.IsTrue(WaitFor(() => rt.IsFastTime));
                SendGrant("run-step-fn", timeStep: 0.5);
                Assert.IsTrue(WaitFor(() => receivedGrant != null && completion != null));
                Assert.AreEqual(0.5, receivedGrant.TimeStep, 0.001);
                Assert.IsFalse(completion.Failed);
                Assert.AreEqual(0, completion.StepNumber);
            } finally {
                controller.Unsubscribe(sub);
                c.Disconnect();
            }
        }

        [Test]
        public void FastTime_GetStepGrant_ReturnsGrant() {
            var c = FreshClient("C# RC FT Wait Test");
            try {
                var rt = new RTIRuntimeControl(c, fastTime: true);
                Thread.Sleep(100);
                ConfigureRun("run-wait", timeStep: 1.0);
                Assert.IsTrue(WaitFor(() => rt.IsFastTime));
                SendGrant("run-wait", timeStep: 1.0);
                var grant = rt.GetStepGrant(timeout: TestTimeoutMs);
                Assert.IsNotNull(grant);
                Assert.AreEqual(1.0, grant.TimeStep, 0.001);
                Assert.AreEqual(0, grant.StepNumber);
                rt.CompleteStep(grant);
            } finally {
                c.Disconnect();
            }
        }

        [Test]
        public void FastTime_Stop_ClearsIsFastTime() {
            var c = FreshClient("C# RC FT Stop Test");
            try {
                var rt = new RTIRuntimeControl(c, fastTime: true);
                Thread.Sleep(100);
                ConfigureRun("run-stop");
                Assert.IsTrue(WaitFor(() => rt.IsFastTime));
                controller.Publish(RTIChannel.RuntimeControl, new RuntimeControl { Stop = new Google.Protobuf.WellKnownTypes.Empty() });
                Assert.IsTrue(WaitFor(() => !rt.IsFastTime));
                Assert.IsFalse(rt.IsFastTime);
            } finally {
                c.Disconnect();
            }
        }

        [Test]
        public void FastTime_GetStepGrant_ReturnsNull_OnStop() {
            var c = FreshClient("C# RC FT Null Test");
            try {
                var rt = new RTIRuntimeControl(c, fastTime: true);
                Thread.Sleep(100);
                ConfigureRun("run-null");
                Assert.IsTrue(WaitFor(() => rt.IsFastTime));
                // Stop while waiting — GetStepGrant must unblock and return null
                new Thread(() => {
                    Thread.Sleep(100);
                    controller.Publish(RTIChannel.RuntimeControl, new RuntimeControl { Stop = new Google.Protobuf.WellKnownTypes.Empty() });
                }) { IsBackground = true }.Start();
                var grant = rt.GetStepGrant(timeout: 2000);
                Assert.IsNull(grant);
            } finally {
                c.Disconnect();
            }
        }

        [Test]
        public void FastTime_ControllerDisconnect_ResetsFastTime() {
            var ctrl = FreshClient("C# RC FT Ctrl Disc Test Ctrl");
            var c = FreshClient("C# RC FT Ctrl Disc Test");
            try {
                var rt = new RTIRuntimeControl(c, fastTime: true);
                Thread.Sleep(100);
                ctrl.Publish(RTIChannel.FastTimeControl, new FastTimeControl {
                    Configure = new FastTimeControl.Types.Configure {
                        ControllerClientId = ctrl.ClientId,
                        RunId = "run-ctrl-disc",
                        TimeStep = 1.0
                    }
                });
                Assert.IsTrue(WaitFor(() => rt.IsFastTime));
                // Disconnect controller while in non-running/paused state (Initial)
                ctrl.Disconnect();
                Assert.IsTrue(WaitFor(() => !rt.IsFastTime, timeoutMs: 3000));
                Assert.IsFalse(rt.IsFastTime);
                Assert.AreEqual(DispatchMode.Immediate, c.DefaultDispatchMode);
            } finally {
                c.Disconnect();
            }
        }

        [Test]
        public void FastTime_ControllerDisconnect_DoesNotReset_WhenRunning() {
            var ctrl = FreshClient("C# RC FT Ctrl Running Test Ctrl");
            var c = FreshClient("C# RC FT Ctrl Running Test");
            try {
                var rt = new RTIRuntimeControl(c, fastTime: true);
                Thread.Sleep(100);
                ctrl.Publish(RTIChannel.FastTimeControl, new FastTimeControl {
                    Configure = new FastTimeControl.Types.Configure {
                        ControllerClientId = ctrl.ClientId,
                        RunId = "run-ctrl-running",
                        TimeStep = 1.0
                    }
                });
                Assert.IsTrue(WaitFor(() => rt.IsFastTime));
                c.State = RuntimeState.Running;
                ctrl.Disconnect();
                Thread.Sleep(300);
                // Fast time should still be active when disconnect happens during Running
                Assert.IsTrue(rt.IsFastTime);
            } finally {
                c.Disconnect();
            }
        }

        // Helper subclass for override hook tests
        private class TestRuntime : RTIRuntimeControl {
            private readonly Action onResetAction;
            private readonly Action onStartAction;
            private readonly Action onStopAction;
            private readonly Action<RuntimeControl.Types.Seek> onSeekAction;
            private readonly bool? loadScenarioResult;

            public TestRuntime(RTIClient rti,
                Action onReset = null, Action onStart = null, Action onStop = null,
                Action<RuntimeControl.Types.Seek> onSeek = null,
                bool? loadScenarioResult = null)
                : base(rti) {
                this.onResetAction = onReset;
                this.onStartAction = onStart;
                this.onStopAction = onStop;
                this.onSeekAction = onSeek;
                this.loadScenarioResult = loadScenarioResult;
            }

            public override void OnReset() => onResetAction?.Invoke();
            public override void OnStart() => onStartAction?.Invoke();
            public override void OnStop() => onStopAction?.Invoke();
            public override void OnSeek(RuntimeControl.Types.Seek seek) => onSeekAction?.Invoke(seek);
            public override bool OnLoadScenario(RuntimeControl.Types.ScenarioSpecification ls, bool playback)
                => loadScenarioResult ?? true;
        }
    }
}
